src/t_zset.c

```diff
@@ -115,15 +115,40 @@ static inline void zslSetNodeHeight(zskiplistNode *x, int height) {
 }
 
 /* Create a skiplist node with the specified number of levels.
- * The SDS string 'ele' is referenced by the node after the call. */
-static zskiplistNode *zslCreateNode(int height, double score, sds ele) {
-    zskiplistNode *zn = zmalloc(sizeof(*zn) + height * sizeof(struct zskiplistLevel));
+ * By embedding elements and levels into the skiplist nodes,
+ * we achieve better cache-friendliness and a more compact memory structure.
+ *
+ * The memory layout is as follows:
+ *
+ *     +-------+------------------+---------+-----+---------+-------------+
+ *     | score | backward-pointer | level-0 | ... | level-N | element-sds |
+ *     +-------+------------------+---------+-----+---------+-------------+
+ */
+static zskiplistNode *zslCreateNode(int height, double score, const_sds ele) {
+    size_t ele_sds_len = ele ? sdslen(ele) : 0;
+    char ele_sds_type = ele ? sdsReqType(ele_sds_len) : 0;
+    size_t ele_sds_size = ele ? sdsReqSize(ele_sds_len, ele_sds_type) : 0;
+    /* Allocate enough space for the node, levels, and the element SDS string.
+     * The +1 is for the header size of the SDS string. */
+    zskiplistNode *zn = zmalloc(sizeof(*zn) + height * sizeof(struct zskiplistLevel) + 1 + ele_sds_size);
     zn->score = score;
-    zn->ele = ele;
     zslSetNodeHeight(zn, height);
+    if (ele) {
+        char *data = ((char *)(zn + 1)) + height * sizeof(struct zskiplistLevel);
+        *data++ = sdsHdrSize(ele_sds_type);
+        sdswrite(data, ele_sds_size, ele_sds_type, ele, ele_sds_len);
+    }
     return zn;
 }
 
+sds zslGetNodeElement(const zskiplistNode *x) {
```

Comment:
Add a line or two documenting the function, like we have for the existing functions. "Returns the embedded element" or something better.

Bts, it takes a `const` node but it returns a non-const sds. Shouldn't it return a `const_sds` instead? (We added the `const_sds` type some year ago.)

Reply:
yeh the downsides of C. I am fine with doing this